sql/parser/walk.go

```diff
@@ -183,13 +189,33 @@ func WalkStmt(v Visitor, stmt Statement) {
 		if stmt.Where != nil {
 			stmt.Where.Expr = WalkExpr(v, stmt.Where.Expr)
 		}
-	case *Insert:
-		switch rows := stmt.Rows.(type) {
-		case Values:
-			for i, tuple := range rows {
-				rows[i] = WalkExpr(v, tuple).(Tuple)
+		for i, expr := range stmt.GroupBy {
+			stmt.GroupBy[i] = WalkExpr(v, expr)
+		}
+		if stmt.Having != nil {
+			stmt.Having.Expr = WalkExpr(v, stmt.Having.Expr)
+		}
+		for i, expr := range stmt.OrderBy {
+			stmt.OrderBy[i].Expr = WalkExpr(v, expr.Expr)
```

Comment:
What we've implemented is what is supported by the postgres grammer.

On Fri, Jul 31, 2015 at 2:38 PM Ben Darnell <notifications@github.com>
wrote:

> In sql/parser/walk.go
> <https://github.com/cockroachdb/cockroach/pull/1880#discussion_r36002629>:
>
> > @@ -183,13 +189,33 @@ func WalkStmt(v Visitor, stmt Statement) {
> >  		if stmt.Where != nil {
> >  			stmt.Where.Expr = WalkExpr(v, stmt.Where.Expr)
> >  		}
> > -	case *Insert:
> > -		switch rows := stmt.Rows.(type) {
> > -		case Values:
> > -			for i, tuple := range rows {
> > -				rows[i] = WalkExpr(v, tuple).(Tuple)
> > +		for i, expr := range stmt.GroupBy {
> > +			stmt.GroupBy[i] = WalkExpr(v, expr)
> > +		}
> > +		if stmt.Having != nil {
> > +			stmt.Having.Expr = WalkExpr(v, stmt.Having.Expr)
> > +		}
> > +		for i, expr := range stmt.OrderBy {
> > +			stmt.OrderBy[i].Expr = WalkExpr(v, expr.Expr)
>
> Many databases do not allow substitution in ORDER BY or other clauses
> referencing columns (on the other hand, users want parameterized ORDER BY,
> and doing this outside the database's placeholder system is a common source
> of SQL injection). The idea is that placeholders are only for values; the
> skeleton of the query should be sufficient for constructing a query plan
> without filling in arguments (so this plan can be constructed once and
> reused for many sets of values).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1880/files#r36002629>.
>